python/ray/data/_internal/execution/streaming_executor.py

```diff
@@ -136,24 +160,40 @@ def _scheduling_loop_step(self, topology: Topology) -> bool:
         # ray.wait() overhead, so make sure to allow multiple dispatch per call for
         # greater parallelism.
         process_completed_tasks(topology)
-        for op_state in topology.values():
-            op_state.refresh_progress_bar()
 
         # Dispatch as many operators as we can for completed tasks.
         limits = self._get_or_refresh_resource_limits()
         cur_usage = self._get_current_usage(topology)
         self._report_current_usage(cur_usage, limits)
-        op = select_operator_to_run(topology, cur_usage, limits)
+        op = select_operator_to_run(
+            topology,
+            cur_usage,
+            limits,
+            ensure_at_least_one_running=self._consumer_blocked(),
+        )
         while op is not None:
             if DEBUG_TRACE_SCHEDULING:
                 _debug_dump_topology(topology)
             topology[op].dispatch_next_task()
             cur_usage = self._get_current_usage(topology)
-            op = select_operator_to_run(topology, cur_usage, limits)
+            op = select_operator_to_run(
+                topology,
+                cur_usage,
+                limits,
+                ensure_at_least_one_running=self._consumer_blocked(),
+            )
+
+        # Update the progress bar to reflect scheduling decisions.
+        for op_state in topology.values():
+            op_state.refresh_progress_bar()
 
         # Keep going until all operators run to completion.
         return not all(op.completed() for op in topology)
 
+    def _consumer_blocked(self) -> bool:
+        """Returns whether the user thread is blocked on topology execution."""
+        return self._runner_thread_out.qsize() == 0
```

Comment:
Isn't this the opposite?

Reply:
When the intra-thread queue is empty, it means the consumer is blocked (blocking get).

Reply:
Ah I see, I read this as "consumer is blocked on other execution". Maybe, "_consumer_idling"?